prototype_source/maskedtensor_adagrad.rst

```diff
@@ -0,0 +1,182 @@
+Efficiency of writing "sparse" semantics for Adagrad
+====================================================
+
+`Issue 1369 <https://github.com/pytorch/pytorch/issues/1369>`__ discussed the additional lines of code
+that were introduce while writing "sparse" semantics for Adagrad.
+But really the code doesn't use sparsity as a compression and optimization technique,
+it wants to use masked semantics. We worked around this by introducing one-off semantics and operators
+that encode this behavior while forcing users to be aware of storage details such as indices and values.
+
+In particular we'll point out when sparsity is used as a semantic extension, i.e. unspecified values are not zero
+and when it is just used to compress zeros.
+We'll also compare and contrast this with equivalent code written using MaskedTensor.
+In the end the code snippets are repeat without additional comments to show the difference in brevity.
+
+Set up
+------
+
+.. code-block:: python
+
+    import torch
+    from torch.masked.maskedtensor import masked_tensor
+
+    # Some hyperparameters
+    eps = 1e-10
+    clr = 0.1
+
+    i = torch.tensor([[0, 1, 1], [2, 0, 2]])
+    v = torch.tensor([3, 4, 5], dtype=torch.float32)
+    grad = torch.sparse_coo_tensor(i, v, [2, 4])
+
+Original sparse implementation
+------------------------------
+
+First, let's break down the current implementation of 
+`Adagrad (functional) <https://github.com/pytorch/pytorch/blob/6c2f235d368b697072699e5ca9485fd97d0b9bcc/torch/optim/_functional.py#L16-L51>`__
+in PyTorch currently:
+
+.. code-block:: python
+    :linenos:
```

Comment:
This may not be your fault, but the line numbers (while nice) are not great:
![image](https://user-images.githubusercontent.com/31798555/192620211-a587a946-30b0-483a-80dc-f572280f998a.png)

Aesthetically, they are too close to the code, but more importantly, one cannot copy/paste the code without also copying the line numbers, which is pretty annoying.

Reply:
yeah - converting to a python file so this won't be an issue anymore